File: android-app/app/src/main/java/com/patienttriage/healthsync/data/ApiModels.kt
package com.patienttriage.healthsync.data

import kotlinx.serialization.Serializable

// Mirrors backend/src/validators/health-data.validator.ts's healthIngestSchema
// exactly — metricType/unit must match backend/src/health/metric-types.ts's
// whitelist, or the server rejects the whole reading. Kept in sync by hand
// since the two projects don't share a schema definition.
@Serializable
data class HealthReading(
    val metricType: String,
    val value: Double,
    val unit: String,
    val recordedAt: String, // ISO-8601, UTC
    val externalId: String,
)

@Serializable
data class IngestRequest(val readings: List<HealthReading>)

@Serializable
data class IngestResponse(
    val connectionId: String,
    val submitted: Int,
    val ingested: Int,
    val skipped: Int,
)

@Serializable
data class LoginRequest(val email: String, val password: String)

@Serializable
data class LoginUser(val id: String, val role: String)

@Serializable
data class LoginResponse(val user: LoginUser, val token: String)

@Serializable
data class HealthConnection(
    val id: String,
    val provider: String,
    val status: String,
    val lastSyncedAt: String? = null,
)

@Serializable
data class ConnectionsResponse(val connections: List<HealthConnection>)

@Serializable
data class ApiErrorBody(val error: ApiErrorDetail)

@Serializable
data class ApiErrorDetail(val message: String)

/** Thrown for any non-2xx response, carrying the server's own error message
 * (backend/src/middleware/error-handler.ts's { error: { message } } shape)
 * where available, so the UI can show something meaningful instead of a
 * raw HTTP status. */
class ApiException(val statusCode: Int, message: String) : Exception(message)

File: ios-app/PatientTriageHealthSync/Models/ApiModels.swift
import Foundation

/// One normalized reading pushed to the shared ingest endpoint — mirrors
/// backend/src/validators/health-data.validator.ts's healthReadingSchema
/// field-for-field, and Android's identical ApiModels.kt HealthReading.
struct HealthReading: Codable {
    let metricType: String
    let value: Double
    let unit: String
    /// ISO-8601 UTC, e.g. "2026-08-22T10:15:00Z".
    let recordedAt: String
    /// HealthKit's own per-sample id (HKSample.uuid) — used for idempotent
    /// re-submission via the backend's (patientId, source, externalRecordId)
    /// unique index.
    let externalId: String
}

struct IngestRequest: Codable {
    let readings: [HealthReading]
}

struct IngestResponse: Codable {
    let connectionId: String
    let submitted: Int
    let ingested: Int
    let skipped: Int
}

struct LoginRequest: Codable {
    let email: String
    let password: String
}

struct LoginUser: Codable {
    let id: String
    let role: String
}

struct LoginResponse: Codable {
    let user: LoginUser
    let token: String
}

struct HealthConnection: Codable {
    let id: String
    let provider: String
    let status: String
    let lastSyncedAt: String?
}

struct ConnectionsResponse: Codable {
    let connections: [HealthConnection]
}

struct ApiErrorDetail: Codable {
    let message: String
}

struct ApiErrorBody: Codable {
    let error: ApiErrorDetail
}

struct ApiError: Error, LocalizedError {
    let statusCode: Int
    let message: String

    var errorDescription: String? { message }
}
